### PR TITLE
DAS-1745 Pt.1 Update all test environments to python 3.9

### PR DESCRIPTION
**Pull request by flamingbear**
## Description

This changes the lockfile strategy to switch from the old explicit version to the new style. Accordingly, that means that the regression test environments are not created with a call to conda-lock rather than conda create.
This new lock strategy allows us to install pip dependencies alongside conda ones.  So this PR also updates the Dockerfile that build the images to use the new strategy, removing the layers that installed harmony-py in the Dockerfile and adding that dependency to each images environment.yml.

This also updates the dependencies where appropriate (places that haven't been updated in a long time) and making sure all tests run under python 3.9

## Jira Issue ID

DAS-1745

## Local Test Steps
Pull this branch, build all images, and run the tests (run_notebooks.sh) against UAT.

## PR Acceptance Checklist
* [ n/a ] Acceptance criteria met
* [ n/a ] Tests added/updated (if needed) and passing
* [ n/a ] Documentation updated (if needed)

### CODE REVIEW

**Review by owenlittlejohns — APPROVED** [review]
I built the new images from the updated `conda-linux-64.lock` files via the `Makefile`. `run_notebooks.sh` successfully ran the `harmony-regression` and `harmony` test suites against UAT. The regression tests in HGA needed to be run against SIT, which has newer code in it (one of the tests expects a newer format for an exception message) - that notebook passed running against Harmony in SIT. Nice work!

**Comment by owenlittlejohns** [comment]
I happened across something that might make the `Dockerfile` (and `environment.yml` file) better. This PR seems like the right time to try it out:

From the [conda documentation](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#create-env-file-manually):

```
name: stats2
channels:
  - javascript
dependencies:
  - python=3.9
  - bokeh=2.4.2
  - conda-forge::numpy=1.21.*
  - nodejs=16.13.*
  - flask
  - pip
  - pip:
    - Flask-Testing
```

It's those bottom three lines that I'm interested in here. These could become (with pinned versions):

```
  - pip
  - pip:
   - harmony-py
```

It seems cleaner than my old bodge fix in the Dockerfile:

```
RUN conda install --name papermill -y -q pip; conda run --name papermill pip install -q harmony-py
```

I'll be playing with this tomorrow as I continue to split the SDS Regression test notebook, but I think this could be a nice improvement.

**Comment by owenlittlejohns** [comment]
Quick update - I don't know if the `harmony-py` bit works. But I think it's the better way to get Pip at least.

**Comment by flamingbear** [comment]
@owenlittlejohns Ok, I think this is ready for re-review. There are a bunch of changes, but I was able to run the images against UAT locally. (well, I'm waiting for SDS to finish, but it's not failed fast)

File: test/Dockerfile [hunk 1]
```diff
@@ -11,17 +11,12 @@ ENV AWS_REGION=us-west-2
...
 
-RUN conda config --add channels conda-forge
-RUN conda install conda-lock -q -y
+RUN conda install --channel=conda-forge --name=base conda-lock -q -y
 
 COPY .netrc .netrc
 RUN mkdir ./${sub_dir}
-COPY ${sub_dir}/conda-linux-64.lock ./${sub_dir}
+COPY ${sub_dir}/conda-lock.yml ./${sub_dir}
+RUN conda-lock install -n papermill ./${sub_dir}/conda-lock.yml
```
Comment (flamingbear) [inline test/Dockerfile:19]: This is the big change in this file. we used to run `conda create`.

**Review by owenlittlejohns — APPROVED** [review]
These changes look great! I have built the images locally and each test suite ran as expected via `run_notebooks.sh`.

Thanks for taking time to work out how to get the `harmony-py` Pip dependency to work with `conda-lock`. I'll mimic this in my pending PR to split the SDS Regression test notebook.

**Pull request merged**